sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb

```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Trace
+      module Export
+        # Implementation of the duck type SpanProcessor that batches spans
+        # exported by the SDK then pushes them to the exporter pipeline.
+        #
+        # All spans reported by the SDK implementation are first added to a
+        # synchronized queue (with a {max_queue_size} maximum size, after the
+        # size is reached spans are dropped) and exported every
+        # schedule_delay_millis to the exporter pipeline in batches of
+        # max_export_batch_size.
+        #
+        # If the queue gets half full a preemptive notification is sent to the
+        # worker thread that exports the spans to wake up and start a new
+        # export cycle.
+        #
+        # max_export_attempts attempts are made to export each batch, while
+        # export fails with {FAILED_RETRYABLE}, backing off linearly in 100ms
+        # increments.
+        class BatchSpanProcessor
+          SCHEDULE_DELAY_MILLIS = 5
+          MAX_QUEUE_SIZE = 2048
+          MAX_EXPORT_BATCH_SIZE = 512
+          MAX_EXPORT_ATTEMPTS = 5
+          private_constant(:SCHEDULE_DELAY_MILLIS, :MAX_QUEUE_SIZE, :MAX_EXPORT_BATCH_SIZE, :MAX_EXPORT_ATTEMPTS)
+
+          def initialize(exporter:,
```

Comment:
Do we expect these settings to come from configuration utlimately? If so, does it make sense to have the constants (above) and default configuration values in addition, or do we only need the defaults in one place?

Reply:
Defaults should live in one place. How do you see configuration working for this?

Reply:
I was wondering if the user would be instantiating the BatchSpanProcessor themselves, or if the SDK would instantiate it on their behalf. If it's the user, then I think things are fine as they are. If it's the SDK, then perhaps we should add options to TraceConfig that default to the values we have in the constants.